### PR TITLE
sparse grid ao along (0,2) axes

### PR DESCRIPTION
**Pull request by blazejba**
When `ao_threshold` arg is specified, it reduces the grid size, if any slices along this axis are all zeros. The reduced grid_size applies to `grid_AO`, `grid_coords`, and `grid_weights`.

### CODE REVIEW

File: notebooks/sparse_grid_ao.ipynb [hunk 1]
```diff
@@ -160,7 +160,7 @@
     "            \n",
     "            for line in output.split(\"\\n\"):\n",
```
Comment (AlexanderMath) [inline notebooks/sparse_grid_ao.ipynb:161]: Did we check if the unit [eV] was correct (or if it should be [meV])? I think it'd be neat to have a cut-off around 42meV (chemical accuracy). If the units are [eV] then we're probably sparsifying a bit too much (150eV error is >1000x worse error than ML models).

File: pyscf_ipu/nanoDFT/nanoDFT.py [hunk 2]
```diff
@@ -87,21 +87,21 @@ def exchange_correlation(density_matrix, grid_AO, grid_weights):
     """Compute exchange correlation integral using atomic orbitals (AO) evalauted on a grid. """
     # Perfectly SIMD parallelizable over grid_size axis.
     # Only need one reduce_sum in the end. 
-    grid_AO_dm = grid_AO[0] @ density_matrix # (gsize, N) @ (N, N) -> (gsize, N)
-    grid_AO_dm = jnp.expand_dims(grid_AO_dm, axis=0) # (1, gsize, N)
+    grid_AO_dm = grid_AO[0] @ density_matrix                                                    # (gsize, N) @ (N, N) -> (gsize, N)
+    grid_AO_dm = jnp.expand_dims(grid_AO_dm, axis=0)                                            # (1, gsize, N)
```
Comment (AlexanderMath) [inline pyscf_ipu/nanoDFT/nanoDFT.py:91]: Amazing we can use the same code for this!

File: pyscf_ipu/nanoDFT/nanoDFT.py [hunk 3]
```diff
@@ -199,10 +199,18 @@ def init_dft_tensors_cpu(mol, opts, DIIS_iters=9):
     grid_weights    = grids.weights                                 # (grid_size,) = (45624,) for C6H6
     coord_str       = 'GTOval_cart_deriv1' if mol.cart else 'GTOval_sph_deriv1'
     grid_AO         = mol.eval_gto(coord_str, grids.coords, 4)      # (4, grid_size, N) = (4, 45624, 9) for C6H6.
-    if opts.ao_threshold is not None:
+    if opts.ao_threshold > 0.0:
```
Comment (AlexanderMath) [inline pyscf_ipu/nanoDFT/nanoDFT.py:202]: Looks good!

Comment (AlexanderMath) [reply pyscf_ipu/nanoDFT/nanoDFT.py:202]: Really like the simplicity: we just pre-process remove and only change grid_size so exchange_correlation code remains the same!

**Review by AlexanderMath — APPROVED** [review]

File: notebooks/sparse_grid_ao.ipynb (hunk 1 above)
Comment (blazejba) [reply notebooks/sparse_grid_ao.ipynb:161]: Yes, the unit was correct. The high error we saw was for 32 carbon atoms placed in a line - the error for placing atoms in a line is high regardless of the grid sparsification (c_32 on the graph). 

image
